README.md
# Password-Generator-CS-Final-Project
This is a program that allows you to generate passwords with
great amounts of flexibility. You have by default four security
levels to choose from, with each one being a little different
from one another. If you are not satisfied you can use the
advanced generator after using the normal one once to
customize your password even greater. After you are done
you can either generate more, quit or save the pws to your
computer.

Instructions:
Steps shown in actual codes

This program has way greater adaptibility and flexibility when generating
passcodes, it also has a save function which allows you to actually
remember it when it's saved as a file on your computer.

The target audience is for people like me who doesn't want to spend
time creating a passcode but at the same time everyone around them
knows their passwords and could easily sabatoge their entire life
by trolling on their accounts if the people want to.




CHANGELOG.md
Date: Friday, 22nd Jan 2022

Author: Michael Zhu

Title: better description

Description: clearly states that option 1 is before 2 in advanced generator

Date: Friday, 22nd Jan 2022

Author: Michael Zhu

Title: bug fix

Description: Fixed a bug where in main.py the input key phrase section wouldn't run after typing continue

Date: Friday, 22nd Jan 2022

Author: Michael Zhu

Title: better comments

Description: Added better explanation to the variables in all files, along with the explanation of program in main.py

Date: Friday, 22nd Jan 2022

Author: Michael Zhu

Title: Refining

Description: Added better explanation to use PwSave.py

Date: Friday, 22nd Jan 2022

Author: Michael Zhu

Title: Bug fixes

Description: Fixed a bug where PwSave.py will only save the key phrase when executed

Date: Friday, 22nd Jan 2022

Author: Michael Zhu

Title: Header

Description: Added header to main.py

Date: Friday, 22nd Jan 2022

Author: Michael Zhu

Title: Comments

Description: Added comments to all files to better explain codes

Date: Saturday, 22nd Jan 2022

Author: Michael Zhu

Title: Formatting

Description: Ran all the files through pep8 to eliminate any formatting mistakes

Date: Friday, 21st Jan 2022

Author: Michael Zhu

Title: Formatting

Description: Ran file AdvancedPwSett.py through pep8 to eliminate any formatting mistakes

Date: Friday, 21st Jan 2022

Author: Michael Zhu

Title: Creation

Description: Imported all the files form Replit


main.py
# Python CS final project - Password Generator
# Michael Zhu
# Jan 20th, 2022
# CS30
# Ms.S

"""This is a program that allows you to generate passwords with
great amounts of flexibility. You have by default four security
levels to choose from, with each one being a little different
from one another. If you are not satisfied you can use the
advanced generator after using the normal one once to
customize your password even greater. After you are done
you can either generate more, quit or save the pws to your
computer"""
# Importing all the modules needed
import string
import random as r
import time
import PwGenProcess as PGen
import AdvancedPwSett as APS
import PwSave

# Setting up the different variables that are going to be used
# within the code
alphabets = list(string.ascii_letters)  # letters word bank

type_count = ["num", "let", "sym"]  # A few lists storing important info
password = []  # the list that is going to hold the password
final_pw_list = []  # the list that holds every pw generated

checkp_1 = 1  # Checkpoints
checkp_2 = 1

# the while loop that ensures the code runs forever till told to stop
while True:
    print("welcome to password generator. This is a place for you to " +
          "develop a satisfactory password with great levels of " +
          "customization. \ntype anything to start, and type quit to " +
          "stop the program\n")
    touch = input()
    if touch == "quit":  # stop
        break

    time.sleep(1)

    # asks for length of password and deals with inproper inputs
    try:
        p_length = int(input("what is the length of your desired password?\n"))
        if p_length < 4:
            print("an accepted password must be at least 4 characters long\n")
            continue
    except ValueError:
        print("please input a number\n")
        continue
        time.sleep(1)

    time.sleep(1)

    # checkpoint 1, keeps looping until an acceptable answer is inputted
    while checkp_1 == 1:
        # asks for the key phrase in password
        must_phrase = input("type in a phrase that you would like to have " +
                            "included in the password, if this doesn't " +
                            "apply to you, press the enter key. Please " +
                            "make sure this phrase is not included in " +
                            "your username for maximium sercurity\n")

        if must_phrase == "\n":
            must_phrase = ""
            checkp_1 -= 1
        elif len(must_phrase) > p_length - 4:
            print("the phrase is too big for the allowed length of " +
                  "password, try again\n")
            continue
        else:
            checkp_1 -= 1

    checkp_1 = 1

    # checkpoint 2, works the same as the previous
    while checkp_2 == 1:
        security_lvl = str(input("What is the level of security you are " +
                                 "looking for?\n\t- low: relatively easier " +
                                 "to memorize, with enough length " +
                                 "also difficult to guess\n\t" +
                                 "- medium: need to write it down, includes " +
                                 "different characters and more randomized" +
                                 "\n\t- high: for the most secure files and " +
                                 "accounts, your must included phrase might" +
                                 "\n\t  be manipulated slightly, " +
                                 "impossible to remember, impossible to " +
                                 "guess\n\t- recommended: a password that " +
                                 "satisfies most websites and account " +
                                 "registeration\n\t  platforms for your " +
                                 "convenience\n"))  # asks for security lvls

        # if improper input happens
        if security_lvl.lower() != "low" and security_lvl.lower() != "med" \
           and security_lvl.lower() != "medium" and security_lvl.lower() != \
           "high" and "recom" not in security_lvl.lower():
            if security_lvl == "quit":
                break
            else:
                print("value not accepted!")
                continue
        else:
            checkp_2 -= 1

    if security_lvl == "quit":  # quits if the improper input is "quit"
        break

    time.sleep(0.5)
    print("generating...")
    time.sleep(1)

    # puts in the key phrase
    password.append(must_phrase)

    # actual generation for password depending on the security level
    # this will run 3 times to generate 3 possible passwords
    for i in range(3):
        result = PGen.GenPro(p_length, must_phrase, type_count, password)

        if security_lvl.lower() == "low":  # low security
            result.low_secur()

        elif security_lvl.lower() == "med" or security_lvl.lower() == "medium":
            result.med_secur()  # medium security

        elif security_lvl.lower() == "high":  # high security
            result.high_secur()

        elif "recom" in security_lvl.lower():  # recommended security
            result.common_website_requirements()

        r.shuffle(result.password)
        print("".join(result.password) + " could be your password")
        time.sleep(0.7)

        # prints out and resets the password list
        final_pw_list.append("".join(result.password))
        password.clear()
        password.append(must_phrase)

    time.sleep(1)

    checkp_2 = 1

    # the end of program but allows other things to be done
    # see line in cont_ask to find out what else could be done
    while True:
        cont_ask = input("do you want to generate more password? type quit " +
                         "to stop the program, continue to use again, " +
                         "advanced to access the advanced generator, or save" +
                         " to save the generated passwords\n")  # things to do

        if "quit" in cont_ask.lower():  # quit
            break
        elif "cont" in cont_ask.lower():  # use again
            break
        elif "adv" in cont_ask.lower():  # use the advanced generator
            time.sleep(1)
            password.clear()
            APS.pw_manipulation(password, final_pw_list)
        elif "sav" in cont_ask.lower():  # save the passwprd
            PwSave.file_save(final_pw_list)
        else:
            print("invalid input!")  # handles wrong inputs

    # breaks or restarts the overall loop
    if "quit" in cont_ask.lower():
        break
    elif "cont" in cont_ask.lower():
        continue


PwSave.py
# importing required modules
import time


"""This is the function that saves the password"""


def file_save(password):

    # setting up required variables
    count = 1  # the number that indicates the sequence of the pws
    checkp_1 = 1  # checkpoint
    save_p = []  # stores all the pws that needs to be saved
    locat = []  # stores the number for saved pws

    time.sleep(1)

    # prints all stored passwords
    for i in password:
        print(f"{count} - {i}")
        count += 1

    time.sleep(1)

    # checkpoint system to accept wrong inputs
    # asks which passwords the users wants to save
    while checkp_1 != 0:
        num_ask = input("which of the following do you want to save? input " +
                        "the number in front of the password. \nSeparate by " +
                        "space and enter all before pressing enter\n")
        num_ask = num_ask.split(" ")

        # handles error
        for i in num_ask:
            try:
                if int(i) > len(password):
                    print("one or more of the index you chose was out of " +
                          "range, please try again\n")
                    break
            except ValueError:
                print("you did not input a number!\n")
                break

        # handles error
        try:
            if int(i) <= len(password):
                checkp_1 -= 1
            else:
                continue
        except ValueError:
            continue

    time.sleep(1)

    # asks what purpose each pw serves to better create the saved file
    for i in num_ask:
        if int(i) <= len(password):
            save_p.append(password[int(i) - 1])
            locat_ask = input("what account or destination is " +
                              f"{password[int(i) - 1]} for?\n\t - ")
            locat.append(locat_ask)

    f = open("generated password.txt", "w+")  # creates a .txt file

    # writes the pws and their purposes in an organized manner in file
    for i in range(len(num_ask)):
        f.write("\t- the password for " + locat[i] + " is " +
                save_p[i])
        f.write("\n----------------------------------------------" +
                "-----------------\n")
    print("password(s) successfully saved\n")


PwGenProcess.py
# importing all required modules
import random as r
import string

# setting up nessesary variables and lists
alphabets = list(string.ascii_letters)  # all the letters
digits = list(string.digits)  # all the numbers
med_spe_characters = list("!@#$%^&*()")  # common symbols
# all symbols
high_spe_characters = list("!@#$%^&*()_-+={[}]|\:;\"\'<,>.?/~`")


"""This is a function that randomizes the upper
and lowercases in your key phrase in high security"""


def randomcase(s):
    result = ''  # right now result doesn't have anything
    for c in s:
        if c in alphabets:  # if c is a letter
            case = r.randint(0, 1)  # randomly chooses between 0 and 1
            if case == 0:
                result += c.upper()
            else:
                result += c.lower()
        else:
            result += c
    return result


"""This class holds the generation codes for all base pw security
levels"""


class GenPro:
    def __init__(self, p_length, must_phrase, type_count, password):
        self.p_length = p_length  # length of password
        self.must_phrase = must_phrase  # key phrase
        self.type_count = type_count  # types of characters in pw
        self.password = password  # the list that stores the result

    def low_secur(self):  # low security
        for i in range(int(self.p_length) - len(self.must_phrase)):
            typpe = r.choice(self.type_count)  # type of characters to add
            if typpe == "num":
                self.password.append(r.choice(digits))
            else:
                self.password.append(r.choice(alphabets))

        for i in range(1, int(self.p_length) - len(self.must_phrase)):
            self.password[i] = self.password[i].lower()

    def med_secur(self):  # medium security
        for i in range(self.p_length - len(self.must_phrase)):
            typpe = r.choice(self.type_count)
            if typpe == "num":
                self.password.append(r.choice(digits))
            elif typpe == "let":
                self.password.append(r.choice(alphabets))
            else:
                self.password.append(r.choice(med_spe_characters))

    def high_secur(self):  # high security
        self.must_phrase = randomcase(self.must_phrase)
        self.password.clear()
        self.password.append(self.must_phrase)

        for i in range(self.p_length - len(self.must_phrase)):
            typpe = r.choice(self.type_count)
            if typpe == "num":
                self.password.append(r.choice(digits))
            elif typpe == "let":
                self.password.append(r.choice(alphabets))
            elif typpe == "sym":
                self.password.append(r.choice(high_spe_characters))

    def common_website_requirements(self):  # recommended security
        x = 0
        for i in range((self.p_length - 4) - len(self.password[0])):
            x = r.randrange(1, 3)
            if x != 2:
                self.password.append(r.choice(digits))
            else:
                self.password.append(r.choice(alphabets))

        # recommended security can satisfy 90% of website
        # registeration by having at least one upper and
        # lower case letters, one number and one symbol
        self.password.append(r.choice(alphabets).lower())
        self.password.append(r.choice(digits))
        self.password.append(r.choice(alphabets).upper())
        self.password.append(r.choice(med_spe_characters))


AdvancedPwSett.py
# Importing the modules
import random as r
import string
import time

# Added the character banks the password could choose from
alphabets = list(string.ascii_letters)  # all the letters
digits = list(string.digits)  # all the numbers
med_spe_characters = list("!@#$%^&*")  # some of the symbols
# all the symbols
high_spe_characters = list("!@#$%^&*()_-+={[}]|\:;\"\'<,>.?/~`")

# Menu for advanced generator
menu = ["select desired character types", "set amount of desired characters",
        "provide word of interest", "prohibit characters",
        "generate password", "quit"]


"""This is the funcion to detect types of characters in passwords"""


def type_detection(charr):
    if charr in alphabets:
        typpe = alphabets
    elif charr in digits:
        typpe = digits
    else:
        typpe = high_spe_characters
    return typpe  # the type of characters


"""The function that executes the advanced generator"""


def pw_manipulation(password, save):

    # Creates variables for detailed questions later on
    checkp_1 = 1
    num_leng_ask = 0
    let_leng_ask = 0
    case_ask = "a"
    sym_leng_ask = 0

    # Greetings
    print("welcome to the advanced settings menu, you can customize your" +
          " password with even\ngreater flexibility in here\n")

    time.sleep(2)

    aska = input("what is your key phrase? if none press enter. " +
                 "Please ensure this phrase does not\nexist in your " +
                 "username\n")  # gets the key phrase

    # If none were inputted key phrase doens't exist
    if aska == "\n":
        aska = ""
    password.append(aska)

    time.sleep(1.5)
    print("Please note that the first two options that will be displayed " +
          "are manetory, choose 1 before 2")
    time.sleep(1.5)

    # The loop that controls the menu
    while True:
        print("You can:\n")
        count = 1  # number before menu option
        for i in menu:  # The previously mentioned menu
            print(str(count) + ". " + i)
            count += 1
            print("\t")

        time.sleep(2)

        # accepts input to choose options from menu
        menu_ask = input("Type 1 for the 1st option, 2 for the 2nd option, " +
                         "and so on, type quit to end the program\n")

        # Select types of characters
        if menu_ask == "1":
            print("you can have:\n\t- numbers\n\t- letters\n\t- symbols\n")
            type_ask = input("which type(s) of characters do you want in " +
                             "your password? type all that apply before " +
                             "pressing enter\n")
            time.sleep(0.5)
            print("Preference(s) noted\n")

        # Enter amount of selected characters
        elif menu_ask == "2":
            # If number in selection
            if "number" in type_ask.lower():
                # get the length of numbers
                num_leng_ask = int(input("how many numbers do you want in " +
                                         "your password?\n"))
                time.sleep(0.5)
                print("understood")

            # If letter in selection
            if "letter" in type_ask.lower():
                # get the length of letters
                let_leng_ask = int(input("how many letters do you want in " +
                                         "your password?\n"))
                time.sleep(0.5)

                # A checkpoint to accept bad inputs and loop it again till
                # Correct input could happen
                while checkp_1 > 0:
                    # Selecting upper, lowercase or both in letter
                    if let_leng_ask < 2:
                        case_ask = input("do you desire upper or lowercase?\n")
                        if "upp" not in case_ask and "low" not in case_ask:
                            print("there's only one character cannot be both")
                            continue
                    else:
                        case_ask = input("do you desire upper, lowercase " +
                                         "or both?")
                    if case_ask.lower() != "both" and "upper" not in \
                       case_ask.lower() and "lower" not in case_ask.lower():
                        print("unaccepted value!")
                    else:
                        print("noted")
                        checkp_1 -= 1

                time.sleep(0.5)

            # If symbol in selection
            if "symbol" in type_ask.lower():
                # get the length of symbols
                sym_leng_ask = int(input("how many symbols do you want in " +
                                         "your password?\n"))
                time.sleep(0.5)
                print("ok")

            # Failsafe
            if "number" not in type_ask.lower() and "letter" not in \
               type_ask.lower() and "symbol" not in type_ask.lower():
                print("You haven't selected your character types yet, " +
                      "please go back and do so\n")

        # Accepts words of interest
        elif menu_ask == "3":
            print("what are your word(s) of interest? This will assist you " +
                  "in better memorization of the\npassword. Separate by " +
                  "space and press enter when finished\n")
            int_phrase_ask = input("")
            int_phrase_ask = int_phrase_ask.split(" ")

        # Prohibits characters
        elif menu_ask == "4":
            print("Type in all the characters you do not want to have in " +
                  "the final product, press enter after you are finished. " +
                  "Note that this will NOT affect your inputted phrases\n")
            proh_ask = input("")

        # Actual generation
        elif menu_ask == "5":
            for i in range(3):  # appends digits
                for i in range(num_leng_ask):
                    password.append(r.choice(digits))
                for i in range(let_leng_ask):  # Appends letters
                    if case_ask.lower() == "upper" or \
                       case_ask.lower() == "uppercase":
                        password.append(r.choice(alphabets).upper())
                    elif "lower" in case_ask.lower():
                        password.append(r.choice(alphabets).lower())
                if case_ask.lower() == "both":
                    password.append(r.choice(alphabets).upper())
                    password.append(r.choice(alphabets).lower())
                    for i in range(let_leng_ask - 2):
                        password.append(r.choice(alphabets))
                for i in range(sym_leng_ask):  # Appends symbols
                    password.append(r.choice(high_spe_characters))

                # Gets rid of prohibited symbols if there are any
                for i in range(1, len(password)):
                    try:
                        if password[i] in proh_ask:
                            typechar = type_detection(password[i])
                            while password[i] in proh_ask:
                                password[i] = r.choice(typechar)
                    except UnboundLocalError:
                        pass

                # Adds words of interest
                try:
                    int_phrase_count = r.randrange(1, 3)
                    for i in range(int_phrase_count):
                        password.append(int_phrase_ask[r.randrange
                                                       (len(int_phrase_ask))])
                except UnboundLocalError:
                    pass

                r.shuffle(password)

                # Prints password
                print("your password could be ", end="")
                for i in password:
                    print(i, end="")
                print("\n")
                save.append("".join(password))
                time.sleep(0.7)

                password.clear()
                password.append(aska)

            time.sleep(2)
            break

        # Quits
        elif menu_ask == "6" or menu_ask == "quit":
            break

        # Accepts wrong inputs
        else:
            print("that is not a menu option...yet")
            time.sleep(1)
